pkg/etcdutil/defrag.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdutil
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// defragData defragments the data directory of each etcd member
+func defragData(tlsConfig *TLSConfig) {
+	client, err := GetTLSClientForEtcd(tlsConfig)
+	if err != nil {
+		logrus.Warnf("failed to create etcd client for defragmentation: %v", err)
+	}
+
+	for _, ep := range tlsConfig.endpoints {
+		logrus.Infof("Defragmenting etcd member[%s]", ep)
+		ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+		_, err := client.Defragment(ctx, ep)
+		cancel()
+		if err != nil {
+			logrus.Warnf("Failed to defragment etcd member[%s] with error: %v", ep, err)
+		} else {
+			logrus.Infof("Finished defragmenting etcd member[%s]", ep)
+		}
+	}
+}
+
+// DefragDataPeriodically defragments the data directory of each etcd member
+func DefragDataPeriodically(stopCh <-chan struct{}, tlsConfig *TLSConfig, d time.Duration) {
```

Comment:
This function is directly invoked as a goroutine in https://github.com/swapnilgm/etcd-backup-restore/blob/defrag/cmd/server.go#L107.

But I don't see any coordination with full and delta snapshots. Shouldn't we pause those before defragmenting? Also, would a full snapshot after defrag make sense?

Reply:
Snapshotting is independent of defragmentation. We won't need to worry about any coordination among them. Etcd takes care of it. At worst, etcd might not respond to snapshot request in case defragmentation takes too much time. And backup-restore sidecar will retry it afterwords.

Reply:
What will the healthcheck response be if snapshot takes too much time and times out?

Reply:
Also, see the comment below regarding coordination during defragmentation.

Reply:
After time out, the health will be set to `false`. And set back to `true` on next successful retry.